### PR TITLE
fix: make sure there is no duplicated rules before updating the netwo…

### LINKED ISSUE #1829: Gain insight into duplicated nsg rules issue
#### What happened:

There is a chance that the PUT NSG request fails because of duplicated rules in the request body, and the reason is still unclear. We can do the following to mitigate:

- [ ] dump the request body if it's failed
- [ ] add a checker before calling the api

#### Environment:

- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by nilo19**
…rk security group

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Make sure there is no duplicated rules before updating the network security group.

#### Which issue(s) this PR fixes:

Fixes #1829

#### Does this PR introduce a user-facing change?

```release-note
fix: make sure there is no duplicated rules before updating the network security group
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure_utils.go [hunk 1]
```diff
@@ -300,3 +300,16 @@ func sameContentInSlices(s1 []string, s2 []string) bool {
...
 }
+
+func removeDuplicatedSecurityRules(rules []network.SecurityRule) []network.SecurityRule {
+	ruleNames := make(map[string]bool)
+	for i := len(rules) - 1; i >= 0; i-- {
+		if _, ok := ruleNames[to.String(rules[i].Name)]; ok {
+			klog.Warningf("Found duplicated rule %s, will be removed.", to.String(rules[i].Name))
+			rules = append(rules[:i], rules[i+1:]...)
+		} else {
+			ruleNames[to.String(rules[i].Name)] = true
```
Comment (lzhecheng) [inline pkg/provider/azure_utils.go:311]: I prefer not put this line in else. Fewer indention.

File: pkg/provider/azure_utils_test.go [hunk 2]
```diff
@@ -255,3 +258,59 @@ func TestGetServiceAdditionalPublicIPs(t *testing.T) {
...
+				{
+					Name: to.StringPtr("rule1"),
+				},
+			},
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			rules := testCase.rules
+			rules = removeDuplicatedSecurityRules(rules)
+			assert.Equal(t, testCase.expected, rules, testCase.description)
```
Comment (lzhecheng) [inline pkg/provider/azure_utils_test.go:313]: I prefer no `testCase.description` here since L310 already includes it.

**Review by lzhecheng — COMMENTED** [review]
A few nits but I won't block its merging.

File: pkg/provider/azure_utils.go (hunk 1 above)
Comment (nilo19) [reply pkg/provider/azure_utils.go]: done

File: pkg/provider/azure_utils_test.go (hunk 2 above)
Comment (nilo19) [reply pkg/provider/azure_utils_test.go]: done

**Pull request merged**